processor/k8sattributesprocessor/processor_test.go

```diff
@@ -1119,6 +1120,33 @@ func TestProcessorAddContainerAttributes(t *testing.T) {
 				conventions.AttributeContainerID:      "5ba4e0e5a5eb1f37bc6e7fc76495914400a3ee309d8828d16407e4b3d5410848",
 			},
 		},
+		{
+			name: "container-imageid",
+			op: func(kp *kubernetesprocessor) {
+				kp.kc.(*fakeClient).Pods[newPodIdentifier("connection", "k8s.pod.ip", "1.1.1.1")] = &kube.Pod{
+					Containers: kube.PodContainers{
+						ByName: map[string]*kube.Container{
+							"app": {
+								Statuses: map[int]kube.ContainerStatus{
+									0: {ImageID: "test/app@sha256:deadbeef00"},
+									1: {ImageID: "test/app@sha256:deadbeef01"},
+									2: {ImageID: "test/app@sha256:deadbeef02"},
+								},
+							},
+						},
+					},
+				}
+			},
+			resourceGens: []generateResourceFunc{
+				withPassthroughIP("1.1.1.1"),
+				withContainerName("app"),
+			},
+			wantAttrs: map[string]string{
+				kube.K8sIPLabelName:                   "1.1.1.1",
+				conventions.AttributeK8SContainerName: "app",
+				containerImageID:                      "test/app@sha256:deadbeef02",
```

Comment:
I think that based on the SemConv spec we have agreed on at https://github.com/open-telemetry/semantic-conventions/blob/main/model/registry/container.yaml, the `container.image.id` should only be the `sha256:deadbeef02` part, which is the container ID that the runtime gives.

If you want to use the full repository digest then you would need to use the [`container.image.repo_digests`](https://github.com/open-telemetry/semantic-conventions/blob/768a23c3627e4bd96146d9f713db214bee7aecf3/model/registry/container.yaml#L60) attribute.

The CRI protocol definition distinguish these 2 at https://github.com/kubernetes/cri-api/blob/c75ef5b473bbe2d0a4fc92f82235efd665ea8e9f/pkg/apis/runtime/v1/api.proto#L1237-L1238 but as far as I can see on k8s we can see both:

```console
$ kubectl get pods --all-namespaces -o jsonpath='{range .items[*]}{"\n"}{.metadata.name}{":\t"}{range .status.containerStatuses[*]}{.imageID}{", "}{end}{end}' |\
sort

coredns-5d78c9869d-29w9c:	sha256:ead0a4a53df89fd173874b46093b6e62d8c72967bbf606d672c9e8c9b601a4fc, 
coredns-5d78c9869d-xkcll:	sha256:ead0a4a53df89fd173874b46093b6e62d8c72967bbf606d672c9e8c9b601a4fc, 
etcd-kind-control-plane:	sha256:86b6af7dd652c1b38118be1c338e9354b33469e69a218f7e290a0ca5304ad681, 
kindnet-885s7:	sha256:b0b1fa0f58c6e932b7f20bf208b2841317a1e8c88cc51b18358310bbd8ec95da, 
kube-apiserver-kind-control-plane:	docker.io/library/import-2023-06-15@sha256:0202953c0b15043ca535e81d97f7062240ae66ea044b24378370d6e577782762, 
kube-controller-manager-kind-control-plane:	docker.io/library/import-2023-06-15@sha256:bdbeb95d8a0820cbc385e44f75ed25799ac8961e952ded26aa2a09b3377dfee7, 
kube-proxy-gr6w4:	docker.io/library/import-2023-06-15@sha256:ce2145a147b3f1fc440ba15eaa91b879ba9cbf929c8dd8f3190868f4373f2183, 
kube-scheduler-kind-control-plane:	docker.io/library/import-2023-06-15@sha256:9d6f903c0d4bf3b145c7bbc68727251ca1abf98aed7f8d2acb9f6a10ac81e8c2,
```

It seems that Kubernetes will either use the `repoDigests` if exists otherwise it will use the `id` to populate the `.status.containerStatuses[*]}{.imageID}`.

On the other hand, from inside a k8s node we can interact directly with the CRI api and inspect the container images directly and verify the differences: 

```console
$ crictl inspecti sha256:a24c7c057ec8730aaa152f77366454835a46dc699fcf243698a622788fd48d62
{
  "status": {
    "id": "sha256:a24c7c057ec8730aaa152f77366454835a46dc699fcf243698a622788fd48d62",
    "repoTags": [
      "registry.k8s.io/metrics-server/metrics-server:v0.7.1"
    ],
    "repoDigests": [
      "registry.k8s.io/metrics-server/metrics-server@sha256:db3800085a0957083930c3932b17580eec652cfb6156a05c0f79c7543e80d17a"
    ],
    "size": "19478031",
     ...
}
```

So I think it's more accurate if we either populate both `container.image.id` and `container.image.repo_digests` conditionally, or we would need to introduce `k8s.container.image.id` which would map exactly whatever the k8s API returns.

Reply:
Originally when I first coded this, I did strip off any prefix and just leave the `sha256:<digest>` portion, and I think this is what most people do want since they can get the image name by using `container.image.name`.

But I read the `container.image.id` "notes" in the semantic conventions which I interpreted to indicate it's acceptable that the value may vary based on the runtime, and it made the code simpler to just propagate the value from the kube api's `PodStatus.ContainerStatuses[].ImageID`.

> K8s defines a link to the container registry repository with digest `"imageID": "registry.azurecr.io /namespace/service/dockerfile@sha256:bdeabd40c3a8a492eaf9e8e44d0ebbb84bac7ee25ac0cf8a7159d25f62555625"`.
The ID is assigned by the container runtime and can vary in different environments. Consider using `oci.manifest.digest` if it is important to identify the same image in different environments/runtimes.

I wasn't involved in any of the semantic convention discussions when these attributes were included though, so I'm happy to try and implement whichever way the semantic convention team thinks is best in this case.

Reply:
> K8s defines a link to the container registry repository with digest "imageID": "registry.azurecr.io /namespace/service/dockerfile@sha256:bdeabd40c3a8a492eaf9e8e44d0ebbb84bac7ee25ac0cf8a7159d25f62555625".

I think this is not 100% accurate since it's not always the case. It seems that when the `repoDigests` of the container image is empty then k8s will use the `id`. So we might need to "fix" the wording there in SemConv.

If we trim the full `repoDigest` to only keep the `sha256:<digest>` part then the remaining is nothing meaningful and definitely not the image's id that the runtime provides.

For example I see the following:

```console
$ kubectl get pods --all-namespaces -o jsonpath='{range .items[*]}{"\n"}{.metadata.name}{":\t"}{range .status.containerStatuses[*]}{.imageID}{", "}{end}{end}' |\
sort
metrics-server-6b4d6cb49f-pfzzs:	registry.k8s.io/metrics-server/metrics-server@sha256:db3800085a0957083930c3932b17580eec652cfb6156a05c0f79c7543e80d17a
```
This means that the `imageID` is the `repoDigest`.

Indeed:

```console
$ crictl inspecti -q registry.k8s.io/metrics-server/metrics-server@sha256:db3800085a0957083930c3932b17580eec652cfb6156a05c0f79c7543e80d17a
{
  "status": {
    "id": "sha256:a24c7c057ec8730aaa152f77366454835a46dc699fcf243698a622788fd48d62",
    "repoTags": [
      "registry.k8s.io/metrics-server/metrics-server:v0.7.1"
    ],
    "repoDigests": [
      "registry.k8s.io/metrics-server/metrics-server@sha256:db3800085a0957083930c3932b17580eec652cfb6156a05c0f79c7543e80d17a"
    ],
    "size": "19478031",
    "uid": {
      "value": "65534"
    },
    "username": "",
    "spec": null,
    "pinned": false
  }
}
```

So if I trim the `registry.k8s.io/metrics-server/metrics-server@sha256:db3800085a0957083930c3932b17580eec652cfb6156a05c0f79c7543e80d17a`:

```console
$ crictl inspecti -q sha256:db3800085a0957083930c3932b17580eec652cfb6156a05c0f79c7543e80d17a
FATA[0000] no such image "sha256:db3800085a0957083930c3932b17580eec652cfb6156a05c0f79c7543e80d17a" present
```

But if I use the actual image id:

```console
$ crictl inspecti -q sha256:a24c7c057ec8730aaa152f77366454835a46dc699fcf243698a622788fd48d62
{
  "status": {
    "id": "sha256:a24c7c057ec8730aaa152f77366454835a46dc699fcf243698a622788fd48d62",
    "repoTags": [
      "registry.k8s.io/metrics-server/metrics-server:v0.7.1"
    ],
    "repoDigests": [
      "registry.k8s.io/metrics-server/metrics-server@sha256:db3800085a0957083930c3932b17580eec652cfb6156a05c0f79c7543e80d17a"
    ],
    "size": "19478031",
    "uid": {
      "value": "65534"
    },
    "username": "",
    "spec": null,
    "pinned": false
  }
}
```

This tells me that Kubernetes either uses the `image.id` or the `image.repoDigests[0]` of the runtime. I haven't checked its implementation though to verify this, maybe it worth checking that.
I believe we should be consistent with the runtime and populate the `container.image.repo_digests` if the `imageID` is a repo digest or use the `container.image.id` if the `imageID` is not repo digest.

Otherwise we need to use a k8s specific attribute called `k8s.container.image.id` for the `imageID` because `container.image.*` is tied to container runtime and we should preserve this.